### PR TITLE
Members only

### PR DESCRIPTION
**Pull request by jairjaramillo**
We created a rails project that allows users to log-in and create posts, which only the logged-in users can see who wrote them.

We used the specifications from this site: https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,33 @@
+# frozen_string_literal: true
+
 class ApplicationController < ActionController::Base
+  def signin(user)
+    new_token = User.create_token
+    user.update_attributes(remember_token: new_token)
+    cookies.permanent[:remember_token] = new_token
+    retrive_user
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: Why not just directly set the `current_user` with the `user`

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,33 @@
...
+    end
+  end
+
+  def current_user
+    @current_user ||= retrive_user
+  end
+
+  def sign_out
+    @current_user = nil
+    cookies[:remember_token] = nil
```
Comment (maelfosso) [inline app/controllers/application_controller.rb]: Just delete it

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  def index
+    @posts = Post.all
+  end
+
+  def logged_in_user
+    return if logged_in?
+
+    flash[:error] = 'Please log in to see this'
+    redirect_to login_path
+  end
```
Comment (maelfosso) [inline app/controllers/posts_controller.rb]: Redirect **unless** there is a connected user

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  before_action :logged_in_user, only: %i[create new]
+
+  def new
+    @post = Post.new
+  end
+
+  def create
+    @post = Post.new(post_params)
+    @user = current_user
+    if @user.posts.build(post_params).save
```
Comment (maelfosso) [inline app/controllers/posts_controller.rb:15]: The `13` and `15` lines it's the same.
One line it's sufficient

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required

Hello @jairjy @uryelah 

Nice introduction to this project, awesome code :smile:
However, I got some issues when testing it

1. Here is what I got when trying to sign up
Screenshot from 2019-12-18 11-46-41
To fix it you need to update this line
https://github.com/jairjy/Microverse-506-Members-Only/blob/a4e52d514703462f31f46f9a9e9797d273fd12cb/config/routes.rb#L13

2. There is also an error when trying to login
Screenshot from 2019-12-18 11-57-46

3. We can access to `http://localhost:3000/posts`
Screenshot from 2019-12-18 12-18-21

You must redirect unless there is a connected user

4. Also read the comments below to know more about the others errors

Kindly make the corresponding changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

**Comment by uryelah** [comment]
@maelfosso Oh no it looks like we crapped up some of the routes when cleaning up unnecessary ones. We'll fix  the issues.
On issue n 4 though:
>We can access to http://localhost:3000/posts
>You must redirect unless there is a connected user

That's not the case according to the Odin project instructions:
>members can see who the author of a post is but, outside, they can only see the story and wonder who wrote it.
...
>Fill out the #index action of the PostsController and create the corresponding view. The view should show a list of every post.
>Now add logic in your Index view to display the author’s name, but only if a user is signed in.

Reference: https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication

View os /posts for not logged user:
[https://i.imgur.com/CboyN3x.png](https://i.imgur.com/CboyN3x.png)
For logged user:
[https://i.imgur.com/oFYspC9.png](https://i.imgur.com/oFYspC9.png)

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (uryelah) [reply app/controllers/application_controller.rb]: Done 👍

File: app/controllers/posts_controller.rb (hunk 4 above)
Comment (uryelah) [reply app/controllers/posts_controller.rb:15]: We actually need the @post variable.
The reason for that is that when you submit a post that's not valid (empty body or title) we want the parts that were already written to still be in the #new view when it's rendered.
If there's no @post variable the render 'new' will  display an empty form regardless of what was first submited.

File: app/controllers/posts_controller.rb (hunk 3 above)
Comment (uryelah) [reply app/controllers/posts_controller.rb]: We changed to use a guard clause due to the rubocop warning form Style/GuardClause best styling practice.
To use unless instead we disabled it around the method.

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -5,7 +5,7 @@ def signin(user)
     new_token = User.create_token
     user.update_attributes(remember_token: new_token)
     cookies.permanent[:remember_token] = new_token
-    retrive_user
+    @current_user = user
```
Comment (samgaco) [inline app/controllers/application_controller.rb:8]: :medal_military:

**Review by samgaco — APPROVED** [review]
Hi @uryelah and @jairjy :high_brightness: 

All the previous requirements have been accomplished and everything pointed out previously is working fine now. The project is approved :+1: 

-----------------------------

* ###### STATUS OF THE PROJECT: APPROVED :heavy_check_mark:

**Pull request merged**